text/0193-fixing-type-unions.md

```diff
@@ -0,0 +1,604 @@
+
+## Generating _union-like_ types
+
+Code generation needs to be changed in `jsii-pacmak` so that languages without native support for _type unions_ (**C#**, **Go**, and **Java**) have
+additional _union-like_ types emitted. The proposed API for a union such as `export type UnionType = Foo | Map<Bar> | Baz[]` in each currently
+supported language without native _type unions_ is the following (subject to change):
+
+- In **C#**
+
+  ```csharp
+  // UnionType.cs
+  public sealed class UnionType {
+    public static UnionType FromFoo(Foo value) { /* ... */ }
+    public static UnionType FromMapOfBar(IReadOnlyDictionary<Bar> value) { /* ... */ }
+    public static UnionType FromListOfBaz(IReadOnlyList<Baz> value) { /* ... */ }
+
+    // Those may throw InvalidCastException
+    public Foo Foo { get; }
+    public IReadOnlyDictionary<Bar> MapOfBar { get; }
+    public IReadOnlyList<Baz> ListOfBaz { get; }
+
+    // Private constructor ensures all initialization happens though the factories
+    private UnionType(/*...*/) { /*...*/ }
+  }
+  ```
+
+- In **Go**:
+
+  > :construction: TBD
+
+- In **Java**:
+
+  ```java
+  // UnionType.java
+  public abstract class UnionType {
+    public static UnionType fromFoo(Foo value) {
+      return new UnionType.FooUnionType(value);
+    }
+    public static UnionType fromMapOfBar(Map<String, Bar> value) {
+      return new UnionType.MapOfBarUnionType(value);
+    }
+    public static UnionType fromListOfBaz(Baz... value) {
+      return fromListOfBaz(Arrays.asList(value));
+    }
+    public static UnionType fromListOfBaz(List<Baz> value) {
+      return new UnionType.ListOfBaz(value);
+    }
+
+    public Foo asFoo() { throw new ClassCastException(); }
+    public Map<String, Bar> asMapOfBar() { throw new ClassCastException(); }
+    public List<Baz> asListOfBaz() { throw new ClassCastException(); }
+
+    // Private constructor ensures all initialization happens through the factories
+    private UnionType(/* ... */) { /* ... */ }
+
+    private static final class FooUnionType extends UnionType { /*...*/ }
+    private static final class MapOfBarUnionType extends UnionType { /*...*/ }
+    private static final class ListOfBazUnionType extends UnionType { /*...*/ }
+  }
+  ```
+
+The _jsii Kernel_ will be aware of the _aliased type unions_, and values passed from **JavaScript** to the _host application_ will be annotated with
+the _union type_'s fully qualified name if the value is not an instance of a known class (this is the case for values of _struct_ types). On the other
+hand, if the value is the instance of a _jsii_ class, the instance will continue to be serialized using this class' fully qualified name, and the
+languages' runtime libraries will wrap that using the _union-like type_ as necessary.
+
+It does not make sense for a _host application_ to use a _type union_ fully qualified name with the _jsii Kernel_'s `create` call: one of the
```

Comment:
If this is true for communication the one way, then it feels like it should be equally true the other way.

Reply:
They can be passed both ways.

The only call you cannot use with is `create`, which will never return a type union either.